include/deal.II/base/vectorization.h

```diff
@@ -3941,6 +3941,211 @@ operator<<(std::ostream &out, const VectorizedArray<Number, width> &p)
+  not_equal             = _CMP_NEQ_OQ,
+  less_than             = _CMP_LT_OQ,
+  less_than_or_equal    = _CMP_LE_OQ,
+  greater_than          = _CMP_GT_OQ,
+  greater_than_or_equal = _CMP_GE_OQ
+};
+
+
+/**
+ * TODO
+ */
+template <SIMDComparison predicate, typename Number>
+DEAL_II_ALWAYS_INLINE inline Number
+compare_and_apply_mask(const Number &left,
+                       const Number &right,
+                       const Number &true_value,
+                       const Number &false_value)
+{
+  bool mask;
+
+  switch (predicate)
+    {
+      case SIMDComparison::equal:
+        mask = (left == right);
+        break;
+      case SIMDComparison::not_equal:
+        mask = (left != right);
+        break;
+      case SIMDComparison::less_than:
+        mask = (left < right);
+        break;
+      case SIMDComparison::less_than_or_equal:
+        mask = (left <= right);
+        break;
+      case SIMDComparison::greater_than:
+        mask = (left > right);
+        break;
+      case SIMDComparison::greater_than_or_equal:
+        mask = (left >= right);
+        break;
+    }
+
+  return mask ? true_value : false_value;
+}
+
+
+/**
+ * TODO
+ */
+template <SIMDComparison predicate, typename Number>
+DEAL_II_ALWAYS_INLINE inline VectorizedArray<Number, 1>
+compare_and_apply_mask(const VectorizedArray<Number, 1> &left,
+                       const VectorizedArray<Number, 1> &right,
+                       const VectorizedArray<Number, 1> &true_value,
+                       const VectorizedArray<Number, 1> &false_value)
+{
+  bool mask;
+
+  switch (predicate)
+    {
+      case SIMDComparison::equal:
+        mask = (left.data == right.data);
+        break;
+      case SIMDComparison::not_equal:
+        mask = (left.data != right.data);
+        break;
+      case SIMDComparison::less_than:
+        mask = (left.data < right.data);
+        break;
+      case SIMDComparison::less_than_or_equal:
+        mask = (left.data <= right.data);
+        break;
+      case SIMDComparison::greater_than:
+        mask = (left.data > right.data);
+        break;
+      case SIMDComparison::greater_than_or_equal:
+        mask = (left.data >= right.data);
+        break;
+    }
+
+  return mask ? true_value.data : false_value.data;
+}
+
+//@}
+
+#ifndef DOXYGEN
+#  if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 3 && defined(__AVX512F__)
+
+template <int predicate>
+DEAL_II_ALWAYS_INLINE inline VectorizedArray<float, 16>
+compare_and_apply_mask(const VectorizedArray<float, 16> &left,
+                       const VectorizedArray<float, 16> &right,
+                       const VectorizedArray<float, 16> &true_values,
+                       const VectorizedArray<float, 16> &false_values)
+{
+  const auto mask = _mm512_cmp_ps(left.data, right.data, predicate);
+  VectorizedArray<float, 16> result;
+  result.data = _mm512_or_ps(_mm512_and_ps(mask, true_values.data),
+                             _mm512_andnot_ps(mask, false_values.data));
```

Comment:
I checked the case for AVX-512. I think we can simply write the function like this:
```C++
const __mmask16 mask = _mm512_cmp_ps(left.data, right.data, predicate);
VectorizedArray<float, 16> result;
result.data = _mm512_mask_mov_ps(false_values.data, mask, true_values.data);
return result;
```
I have not verified it yet, but this seems to be even simpler than the SSE/AVX case.